File: extensions/eclipse-runlens/src/runlenseclipse/DaemonClient.java
package runlenseclipse;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.WebSocket;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.atomic.AtomicInteger;

import com.google.gson.Gson;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

public class DaemonClient {
    private final String url;
    private final HttpClient client = HttpClient.newHttpClient();
    private final Gson gson = new Gson();
    private final AtomicInteger nextId = new AtomicInteger(1);
    private final ConcurrentHashMap<Integer, PendingRequest> pending = new ConcurrentHashMap<>();
    private final CopyOnWriteArrayList<DaemonListener> listeners = new CopyOnWriteArrayList<>();
    private volatile WebSocket ws;

    public interface DaemonListener {
        default void onConnected() {}
        default void onDisconnected() {}
        default void onEvent(String method, JsonObject params) {}
    }

    private static class PendingRequest {
        final CompletableFuture<JsonElement> future;
        PendingRequest(CompletableFuture<JsonElement> future) { this.future = future; }
    }

    public DaemonClient(String url) { this.url = url; }

    public CompletableFuture<Void> connect() {
        CompletableFuture<Void> future = new CompletableFuture<>();
        client.newWebSocketBuilder()
            .buildAsync(URI.create(url), new WebSocket.Listener() {
                @Override
                public void onOpen(WebSocket webSocket) {
                    ws = webSocket;
                    listeners.forEach(DaemonListener::onConnected);
                    future.complete(null);
                }

                @Override
                public void onText(WebSocket webSocket, CharSequence data, boolean last) {
                    String msg = data.toString();
                    JsonObject json = JsonParser.parseString(msg).getAsJsonObject();
                    Integer id = json.has("id") ? json.get("id").getAsInt() : null;
                    if (id != null) {
                        PendingRequest pr = pending.remove(id);
                        if (pr != null) {
                            if (json.has("error")) {
                                pr.future.completeExceptionally(
                                    new RuntimeException(json.get("error").toString()));
                            } else {
                                pr.future.complete(json.get("result"));
                            }
                        }
                    } else {
                        String method = json.has("method") ? json.get("method").getAsString() : "";
                        JsonObject params = json.has("params") ? json.getAsJsonObject("params") : new JsonObject();
                        for (DaemonListener l : listeners) l.onEvent(method, params);
                    }
                }

                @Override
                public void onError(WebSocket webSocket, Throwable error) {
                    future.completeExceptionally(error);
                }

                @Override
                public void onClose(WebSocket webSocket, int statusCode, String reason) {
                    ws = null;
                    listeners.forEach(DaemonListener::onDisconnected);
                }
            });
        return future;
    }

    public void disconnect() {
        if (ws != null) ws.sendClose(1000, "client shutdown");
        ws = null;
    }

    public CompletableFuture<JsonElement> call(String method) {
        return call(method, Map.of());
    }

    public CompletableFuture<JsonElement> call(String method, Object params) {
        int id = nextId.getAndIncrement();
        CompletableFuture<JsonElement> future = new CompletableFuture<>();
        pending.put(id, new PendingRequest(future));
        Map<String, Object> req = Map.of(
            "jsonrpc", "2.0", "id", id, "method", method, "params", params);
        try {
            ws.sendText(gson.toJson(req), true);
        } catch (Exception e) {
            pending.remove(id);
            future.completeExceptionally(e);
        }
        return future;
    }

    public void addListener(DaemonListener listener) { listeners.add(listener); }
    public void removeListener(DaemonListener listener) { listeners.remove(listener); }
}

File: extensions/jetbrains-runlens/src/main/kotlin/runlens/DaemonClient.kt
package runlens

import com.google.gson.Gson
import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import java.net.URI
import java.net.http.HttpClient
import java.net.http.WebSocket
import java.util.concurrent.CancellationException
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CopyOnWriteArrayList

class DaemonClient(private val url: String = "ws://localhost:9876") {
    private val gson = Gson()
    private var ws: WebSocket? = null
    private val pending = ConcurrentHashMap<Int, PendingRequest>()
    private var nextId = 1
    private val listeners = CopyOnWriteArrayList<DaemonListener>()
    private var destroyed = false

    data class PendingRequest(val resolve: (JsonElement?) -> Unit, val reject: (Throwable) -> Unit)

    interface DaemonListener {
        fun onConnected() {}
        fun onDisconnected() {}
        fun onEvent(method: String, params: JsonObject) {}
    }

    fun connect(): CompletableFuture<Void> {
        val future = CompletableFuture<Void>()
        destroyed = false
        val client = HttpClient.newBuilder().build()
        client.newWebSocketBuilder()
            .buildAsync(URI.create(url), object : WebSocket.Listener {
                override fun onOpen(webSocket: WebSocket) {
                    ws = webSocket
                    listeners.forEach { it.onConnected() }
                    future.complete(null)
                    WebSocket.Listener.super.onOpen(webSocket)
                }

                override fun onText(webSocket: WebSocket, data: CharSequence, last: Boolean) {
                    if (destroyed) return
                    val message = data.toString()
                    val json = JsonParser.parseString(message).asJsonObject
                    val id = json.get("id")?.asInt
                    if (id != null) {
                        val pending = pending.remove(id) ?: return
                        val error = json.get("error")
                        if (error != null) {
                            pending.reject(RuntimeException(error.toString()))
                        } else {
                            pending.resolve(json.get("result"))
                        }
                    } else {
                        listeners.forEach { it.onEvent(
                            json.get("method")?.asString ?: "",
                            json.getAsJsonObject("params") ?: JsonObject()
                        ) }
                    }
                }

                override fun onError(webSocket: WebSocket, error: Throwable) {
                    if (!destroyed) future.completeExceptionally(error)
                }

                override fun onClose(webSocket: WebSocket, statusCode: Int, reason: String) {
                    ws = null
                    listeners.forEach { it.onDisconnected() }
                }
            })
        return future
    }

    fun disconnect() {
        destroyed = true
        ws?.sendClose(1000, "client shutdown")
        ws = null
    }

    fun call(method: String, params: Any = mapOf<String, Any>()): CompletableFuture<JsonElement?> {
        val id = nextId++
        val future = CompletableFuture<JsonElement?>()
        pending[id] = PendingRequest({ future.complete(it) }, { future.completeExceptionally(it) })
        val request = mapOf(
            "jsonrpc" to "2.0",
            "id" to id,
            "method" to method,
            "params" to params
        )
        try {
            ws?.sendText(gson.toJson(request), true)
        } catch (e: Exception) {
            pending.remove(id)
            future.completeExceptionally(e)
        }
        return future
    }

    fun addListener(listener: DaemonListener) {
        listeners.add(listener)
    }

    fun removeListener(listener: DaemonListener) {
        listeners.remove(listener)
    }
}